### PR TITLE
[1.0.0] Dynamic Framework Creation on Compile Time and Run Phase Integration

### PR DESCRIPTION
**Pull request by rweindl**
@pkluz @paytonmiller @kevgo This PR tracks the progress on `OriginateTheme` version 1.0.0. The release of the new version is implementing a new concept of dynamically creating access to themes  provided in the `JSON` files. After integrating the OriginateTheme framework into a project the specified `JSON` file will automatically get parsed and via a code generation phase transformed to accessible Objective-C classes. These created or modified classes are added to the OriginateTheme framework before each source code compilation.

In addition, it is possible to override theme styles dynamically on runtime by creating a new instance of the `OTTheme` object and providing a path to a `JSON` file which overwrites (partially) the at compile time defined styles will be used as fallback. This will allow (remote) customization of the OriginateTheme framework also after submitting or distributing the application.

The implementation of the framework is finished, an example project is added, all files are in a state I would submit them for review and a comprehensive documentation inside README file is added. If you want you can already start giving me a review on these files.

Since I still have to implement tests this PR remains with a work in progress label.

Before this branch is merged on master the Podfile in the example folder has to be updated to point to the new created version tag of the .podspec or the Originate .podspec.

### CODE REVIEW

File: OriginateTheme/Sources/Classes/OTTheme.m [hunk 1]
```diff
@@ -1,25 +1,28 @@
...
+@property (nonatomic, strong, readwrite) OTColors *colors;
+@property (nonatomic, strong, readwrite) OTComponents *components;
+@property (nonatomic, strong, readwrite) OTFonts *fonts;
 
 @end
 
-@implementation OriginateTheme
+@implementation OTTheme
 
 #pragma mark - OriginateUI
```
Comment (pkluz) [inline OriginateTheme/Sources/Classes/OTTheme.m]: -> `OTTheme`

File: Example/OriginateTheme-Example/OTERemoteThemeViewController.m [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+              return;
+          }
+          
+          if (!data) {
+              NSLog(@"Received no data!");
+              return;
+          }
+          
+          dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
+
```
Comment (pkluz) [inline Example/OriginateTheme-Example/OTERemoteThemeViewController.m]: Stray empty space

File: Example/OriginateTheme-Example/OTERemoteThemeViewController.m [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+          
+          if (!data) {
+              NSLog(@"Received no data!");
+              return;
+          }
+          
+          dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
+              
+              NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
+              NSString *documentsDirectory = [paths objectAtIndex:0];
```
Comment (pkluz) [inline Example/OriginateTheme-Example/OTERemoteThemeViewController.m]: Use `firstObject`

File: OriginateTheme/Scripts/ot_generator.py [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+
+def headerTemplate():
+    """
+        Template for an OriginateTheme header file.
+    """
+    return """//
+//  $OriginateThemeClassName.h
+//  OriginateTheme
+//
+//  Created by Robert Weindl on 2016-09-13.
```
Comment (pkluz) [inline OriginateTheme/Scripts/ot_generator.py]: Let's remove the created by line in all templates as they get autogenerated

File: OriginateTheme/Sources/Categories/UIColor+OTHex.m [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
 
-@implementation UIColor (OriginateThemeHex)
+@implementation UIColor (OTHex)
 
-+ (instancetype)oth_colorWithHexString:(NSString *)hexString
+#pragma mark - UIColor (OTHex)
+
++ (instancetype)ot_colorWithHexString:(NSString *)hexString
 {
     if ([hexString length] != 6) {
```
Comment (pkluz) [inline OriginateTheme/Sources/Categories/UIColor+OTHex.m]: If you could, please extend this to also allow for RGBA (8 chars wherein the last two are alpha value)

**Comment by pkluz** [comment]
Please make sure - after merging - to add a PR to this repository: https://github.com/Originate/CocoaPods/tree/master/OriginateTheme exposing the Podspec there as well.

Please add two github issues to this project, namely:
- [ ] Break the python script into smaller pieces by extracting the code-snippets into proper template files.
- [ ] Add unit tests for the code generation (given static input A expect answer B). Python has `unittest` for this, it should totally suffice.

**Comment by rweindl** [comment]
@pkluz Please take a look and approve the created unit tests by executing from the OriginateTheme directory the following command:
```
python -m unittest discover -s OriginateTheme/Scripts -p "*_test.py"
```

Gonna setup Travis right after approval.

**Comment by pkluz** [comment]
Lookin good!~

**Pull request closed without merging**